Bb8 pool

to266 opened the pull request:
https://github.com/eventually-rs/eventually-rs/issues/136

Once I manage to run the tests (hopefully in the CI), I'll have a look if the subscriptions actually should use a connection that is not managed by the pool, as hinted at here: [dedicated_connection](https://docs.rs/bb8/0.4.2/bb8/struct.Pool.html#method.dedicated_connection)

to266 commented:
Another idea would be to make the connection pool inside the `eventually-postgres` into `pub(crate) pool`, which would allow not passing it everywhere. I think it makes somewhat more sense to do that, since it replaced a single connection.

patch would look like this:
```patch
diff --git a/eventually-postgres/src/store.rs b/eventually-postgres/src/store.rs
index cedb314..0153bdb 100644
--- a/eventually-postgres/src/store.rs
+++ b/eventually-postgres/src/store.rs
@@ -239,7 +239,7 @@ where
     <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
 {
     pub(crate) type_name: &'static str,
-    pool: Pool<PostgresConnectionManager<Tls>>,
+    pub(crate) pool: Pool<PostgresConnectionManager<Tls>>,
     id: std::marker::PhantomData<Id>,
     payload: std::marker::PhantomData<Event>,
 }
diff --git a/eventually-postgres/src/subscription.rs b/eventually-postgres/src/subscription.rs
index 60303e4..0d9a027 100644
--- a/eventually-postgres/src/subscription.rs
+++ b/eventually-postgres/src/subscription.rs
@@ -13,8 +13,7 @@ use futures::stream::{StreamExt, TryStreamExt};
 
 use serde::{Deserialize, Serialize};
 
-use bb8::{Pool, RunError};
-use bb8_postgres::PostgresConnectionManager;
+use bb8::RunError;
 use tokio_postgres::tls::{MakeTlsConnect, TlsConnect};
 use tokio_postgres::Socket;
 
@@ -65,7 +64,6 @@ where
     <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
     <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
 {
-    pool: Pool<PostgresConnectionManager<Tls>>,
     store: EventStore<SourceId, Event, Tls>,
     subscriber: EventSubscriber<SourceId, Event>,
 }
@@ -83,12 +81,10 @@ where
     ///
     /// [`PersistentBuilder`]: struct.PersistentBuilder.html
     pub fn new(
-        pool: Pool<PostgresConnectionManager<Tls>>,
         store: EventStore<SourceId, Event, Tls>,
         subscriber: EventSubscriber<SourceId, Event>,
     ) -> Self {
         Self {
-            pool,
             store,
             subscriber,
         }
@@ -108,7 +104,7 @@ where
     ) -> Result<Persistent<SourceId, Event, Tls>, RunError<tokio_postgres::Error>> {
         let params: Params = &[&name, &self.store.type_name];
 
-        let client = self.pool.get().await?;
+        let client = self.store.pool.get().await?;
         let row = client.query_one(GET_OR_CREATE_SUBSCRIPTION, params).await?;
 
         let last_sequence_number: i64 = row.try_get("last_sequence_number")?;
@@ -116,7 +112,6 @@ where
         Ok(Persistent {
             name,
             last_sequence_number: AtomicI64::from(last_sequence_number),
-            pool: self.pool.clone(),
             store: self.store.clone(),
             subscriber: self.subscriber.clone(),
         })
@@ -139,7 +134,6 @@ where
 {
     name: String,
     last_sequence_number: AtomicI64,
-    pool: Pool<PostgresConnectionManager<Tls>>,
     store: EventStore<SourceId, Event, Tls>,
     subscriber: EventSubscriber<SourceId, Event>,
 }
@@ -244,7 +238,7 @@ where
                 "Checkpointing persistent subscription"
             );
 
-            let client = self.pool.get().await.map_err(Error::Checkpoint)?;
+            let client = self.store.pool.get().await.map_err(Error::Checkpoint)?;
             client
                 .execute(CHECKPOINT_SUBSCRIPTION, params)
                 .await
diff --git a/eventually-postgres/tests/subscriber.rs b/eventually-postgres/tests/subscriber.rs
index f9663f8..2ceaf4e 100644
--- a/eventually-postgres/tests/subscriber.rs
+++ b/eventually-postgres/tests/subscriber.rs
@@ -122,7 +122,7 @@ async fn persistent_subscription_works() {
         .await
         .expect("failed to create event subscription");
 
-    let subscription = PersistentBuilder::new(pool, event_store.clone(), event_subscriber)
+    let subscription = PersistentBuilder::new(event_store.clone(), event_subscriber)
         .get_or_create(source_name.to_owned())
         .await
         .expect("failed to create persistent subscription");

```

ar3s3ru commented:
Hey @to266,

Thanks for submitting the PR! :tada: 

Couple of things before I go into full-review mode:

>if the subscriptions actually should use a connection that is not managed by the pool

The `PersistentSubscription` **should** use the connection pool for its operations (checkpointing). But the `EventSubscriber` needs a dedicated connection, due to its usage of `NOTIFY/LISTEN` for pub/sub behavior.

>Another idea would be to make the connection pool inside the eventually-postgres into pub(crate) pool, which would allow not passing it everywhere. I think it makes somewhat more sense to do that, since it replaced a single connection.

I have somewhat mixed feelings about that, as `PersistentSubscription` would now depend from the internal structure of `EventStore`, which is going to make future refactorings a bit more complicated I'm afraid. We could argue that `PersistentSubscription` already depends on `EventStore`, but only on its public API -- whose changes would affect way more than just the `Subscription`, but likely the whole crate and users and implementors :sweat_smile: 

Haven't picked my mind yet, what do you think? @to266

to266 commented:
Sure, don't mind passing the pool explicitly - it's not too bothersome, and your reasons are sound as far as I can tell.

Let me see if I can figure out the dedicated connection for `EventSubscriber`

to266 commented:
Hmm, ok, as far as I can tell, giving `pool` to the `EventSubscriber` is not really useful, as the dedicated connection needs access not only to the `Client` (which is not a problem), but also to the `Connection` - and this one is hidden away by the pool.

So for now I think it might be best to not touch that part, and create subscribers from DSN as it is. I guess this leaves them brittle when compared to other parts, but it also feels out of scope of the PR

ar3s3ru commented:
>So for now I think it might be best to not touch that part, and create subscribers from DSN as it is. I guess this leaves them brittle when compared to other parts, but it also feels out of scope of the PR

Consider this: subscriptions are used for running projections. Usually projections are run in a different pod/execution environment than the write API. Persisted subscriptions are also stateful, meaning that they will track the last event processed.

So, even if the connection between the projection pod and the database is lost, the pod will most likely restart (due to an error with the connection) and will connect to the new database IP, resuming from where it left off.

Of course, there's gonna be eventual consistency in there. But if you're using projections, you would've made peace with eventual consistency already by then :smile:

ar3s3ru commented:
So @to266, is the PR ready for review? :eyes:

to266 commented:
Yes!

ar3s3ru commented on eventually-postgres/src/store.rs:
```diff
@@ -121,11 +149,23 @@ impl EventStoreBuilder {
-pub struct EventStoreBuilderMigrated {
-    client: Arc<Client>,
+pub struct EventStoreBuilderMigrated<Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
```
Do you need these clauses here, if they're specified again in the `impl<Tls> EventStoreBuilderMigrated<Tls>` block?

to266 replied:
Yeah, sadly it seems I do. I'm not too happy with the super long trait bounds list, I wish there was a way to create an alias for that.. But if I remove any right now, there are many complaints

to266 replied:
https://doc.rust-lang.org/unstable-book/language-features/trait-alias.html Seems to be at least on the way there (though not sure it even allows replacing all four lines)

ar3s3ru replied:
Oh yeah indeed, it seems the trait bound is defined on the manager structure itself: https://docs.rs/bb8-postgres/0.4.0/src/bb8_postgres/lib.rs.html#18-20

ar3s3ru commented on eventually-postgres/src/store.rs:
```diff
@@ -136,9 +176,9 @@ impl EventStoreBuilderMigrated {
         &self,
         type_name: &'static str,
-    ) -> std::result::Result<EventStore<Id, Event>, tokio_postgres::Error> {
+    ) -> std::result::Result<EventStore<Id, Event, Tls>, RunError<tokio_postgres::Error>> {
```
What do you think about defining a type alias?

```rust
pub type PoolResult<T> = std::result::Result<T, RunError<tokio_postgres::Error>>
```

So you can change this statement to:
```rust
pub async fn build<Id, Event>(
    &self,
    type_name: &'static str,
) -> PoolResult<EventStore<Id, Event, Tls>>
```

Same goes for this: https://github.com/eventually-rs/eventually-rs/pull/137/files#diff-d419ac38248513a9a51b12671bf15f4dbf69d2b16002da96ae05f765edad9657R204-R207

ar3s3ru commented on eventually-postgres/src/store.rs:
```diff
@@ -188,21 +231,31 @@ impl EventStoreBuilderMigrated {
 #[derive(Debug, Clone)]
-pub struct EventStore<Id, Event> {
+pub struct EventStore<Id, Event, Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
```
Do you need to specify these constraints here? Given the `impl` block already has them.

ar3s3ru commented on eventually-postgres/src/store.rs:
```diff
@@ -337,26 +397,33 @@ impl<Id, Event> EventStore<Id, Event> {
-    async fn create_aggregate_type(&self) -> std::result::Result<(), tokio_postgres::Error> {
+    async fn create_aggregate_type(
+        &self,
+    ) -> std::result::Result<(), RunError<tokio_postgres::Error>> {
```
If you define the `PoolResult` alias, you can simplify this return type :+1:

ar3s3ru commented on eventually-postgres/src/subscription.rs:
```diff
@@ -50,33 +52,43 @@ pub enum Error {
-    client: Arc<Client>,
-    store: EventStore<SourceId, Event>,
+pub struct PersistentBuilder<SourceId, Event, Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
```
Do you need these constraints here?

ar3s3ru reviewed (commented):
Great job :100: 

Left a couple of comments. Another point is: can you also update the crate documentation? https://github.com/eventually-rs/eventually-rs/blob/02dc01e290136e405864a9501fdcb4084ecf9060/eventually-postgres/src/lib.rs#L15-L39

to266 commented:
Updated the docs, refactored the `PoolResult`, and did not manage to remove the trait bounds without unhappy compiler. Thanks for the review, please have a look again when you find the chance! :+1:

to266 commented:
Anything else left to be done here @ar3s3ru? From my point if new this is done right now :)
